python/tank_vendor/shotgun_deploy/import_handler.py

```diff
@@ -0,0 +1,323 @@
+# Copyright (c) 2016 Shotgun Software Inc.
+#
+# CONFIDENTIAL AND PROPRIETARY
+#
+# This work is provided "AS IS" and subject to the Shotgun Pipeline Toolkit
+# Source Code License included in this distribution package. See LICENSE.
+# By accessing, using, copying or modifying this work you indicate your
+# agreement to the Shotgun Pipeline Toolkit Source Code License. All rights
+# not expressly granted therein are reserved by Shotgun Software Inc.
+
+import imp
+import os
+import sys
+
+
+class CoreImportHandler(object):
+    """A custom import handler to allow for core version switching.
+
+    Usage:
+        >>> import sys
+        >>> from tank.shotgun_deploy import CoreImportHandler
+        >>> importer = CoreImportHandler("/path/to/a/version/of/tk-core")
+        >>> sys.meta_path.append(importer)
+        >>> # import/run a bunch of code
+        >>> # context change, need to use a different core
+        >>> importer.set_core_path("/path/to/a/different/version/of/tk-core")
+        >>> # namespaces cleared. new imports will come from new core location
+
+    When an instance of this object is added to `sys.meta_path`, it is used to
+    alter the way python imports packages.
+
+    The core path is used to locate modules attempting to be loaded. The core
+    path can be set via `set_core_path` to alter the location of existing and
+    future core imports.
+
+    For more information on custom import hooks, see PEP 302:
+        https://www.python.org/dev/peps/pep-0302/
+
+    """
+
+    def __init__(self, core_path):
+        """Initialize the custom importer.
+
+        :param core_path: A str path to the core location to import from.
+
+        """
+
+        self._core_path = None
+        self._namespaces = []
+
+        # a dictionary to hold module information after it is found, before
+        # it is loaded.
+        self._module_info = {}
+
+        # re-imports any existing modules for the core namespaces
+        self.set_core_path(core_path)
```

Comment:
My personal preference is to do it this way since the path is required for the handler to be useful. I could be convinced otherwise if you feel strongly. :)